File: ppd/contracts/source_index.py
"""PP&D source-index record contracts."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Optional


class CrawlStatus(str, Enum):
    FETCHED = "fetched"
    NOT_MODIFIED = "not_modified"
    SKIPPED = "skipped"
    FAILED = "failed"


class SourcePageType(str, Enum):
    GUIDANCE = "guidance"
    FORM_INDEX = "form_index"
    FAQ = "faq"
    PORTAL_REFERENCE = "portal_reference"
    PUBLIC_SEARCH_REFERENCE = "public_search_reference"
    PDF = "pdf"
    OTHER = "other"


@dataclass(frozen=True)
class RedirectHop:
    from_url: str
    to_url: str
    status_code: int
    observed_at: str

    def validate(self) -> list[str]:
        errors: list[str] = []
        if not self.from_url.startswith("https://"):
            errors.append("redirect from_url must be HTTPS")
        if not self.to_url.startswith("https://"):
            errors.append("redirect to_url must be HTTPS")
        if self.status_code not in {301, 302, 303, 307, 308}:
            errors.append("redirect status_code must be an HTTP redirect code")
        if not self.observed_at.endswith("Z"):
            errors.append("redirect observed_at must end in Z")
        return errors


@dataclass(frozen=True)
class SourceIndexRecord:
    id: str
    source_url: str
    canonical_url: str
    title: str
    bureau: str
    page_type: SourcePageType
    content_type: str
    first_seen_at: str
    last_seen_at: str
    crawl_status: CrawlStatus
    redirects: tuple[RedirectHop, ...] = field(default_factory=tuple)
    fetched_at: Optional[str] = None
    http_status: Optional[int] = None
    etag: Optional[str] = None
    last_modified: Optional[str] = None
    content_hash: Optional[str] = None
    skip_reason: Optional[str] = None
    failure_reason: Optional[str] = None

    def validate(self) -> list[str]:
        errors: list[str] = []
        if not self.id.strip():
            errors.append("id is required")
        if not self.source_url.startswith("https://"):
            errors.append("source_url must be HTTPS")
        if not self.canonical_url.startswith("https://"):
            errors.append("canonical_url must be HTTPS")
        if not self.title.strip():
            errors.append("title is required")
        if not self.bureau.strip():
            errors.append("bureau is required")
        if "/" not in self.content_type:
            errors.append("content_type must be MIME-like")
        if not self.first_seen_at.endswith("Z"):
            errors.append("first_seen_at must end in Z")
        if not self.last_seen_at.endswith("Z"):
            errors.append("last_seen_at must end in Z")
        if self.fetched_at is not None and not self.fetched_at.endswith("Z"):
            errors.append("fetched_at must end in Z")

        for redirect in self.redirects:
            errors.extend(redirect.validate())
        if self.redirects:
            if self.redirects[0].from_url != self.source_url:
                errors.append("first redirect must start at source_url")
            if self.redirects[-1].to_url != self.canonical_url:
                errors.append("last redirect must end at canonical_url")
        elif self.source_url != self.canonical_url:
            errors.append("redirects are required when source_url differs from canonical_url")

        if self.crawl_status in {CrawlStatus.FETCHED, CrawlStatus.NOT_MODIFIED}:
            if not self.fetched_at:
                errors.append(f"{self.crawl_status.value} records require fetched_at")
            if self.http_status not in {200, 304}:
                errors.append(f"{self.crawl_status.value} records require http_status 200 or 304")
            if not (self.content_hash or "").startswith("sha256:"):
                errors.append(f"{self.crawl_status.value} records require sha256 content_hash")
        if self.crawl_status == CrawlStatus.NOT_MODIFIED and self.http_status != 304:
            errors.append("not_modified records require http_status 304")
        if self.crawl_status == CrawlStatus.SKIPPED and not (self.skip_reason or "").strip():
            errors.append("skipped records require skip_reason")
        if self.crawl_status == CrawlStatus.FAILED and not (self.failure_reason or "").strip():
            errors.append("failed records require failure_reason")
        return errors

File: ppd/contracts/source_index.ts
export type CrawlStatus = "fetched" | "not_modified" | "skipped" | "failed";

export type SourcePageType =
  | "guidance"
  | "form_index"
  | "faq"
  | "portal_reference"
  | "public_search_reference"
  | "pdf"
  | "other";

export interface RedirectHop {
  fromUrl: string;
  toUrl: string;
  statusCode: 301 | 302 | 303 | 307 | 308;
  observedAt: string;
}

export interface SourceIndexRecord {
  id: string;
  sourceUrl: string;
  canonicalUrl: string;
  title: string;
  bureau: string;
  pageType: SourcePageType;
  contentType: string;
  firstSeenAt: string;
  lastSeenAt: string;
  crawlStatus: CrawlStatus;
  redirects: RedirectHop[];
  fetchedAt?: string;
  httpStatus?: number;
  etag?: string;
  lastModified?: string;
  contentHash?: string;
  skipReason?: string;
  failureReason?: string;
}

const REDIRECT_CODES = new Set([301, 302, 303, 307, 308]);
const HASH_RE = /^sha256:[0-9a-f]{64}$/;

export function validateSourceIndexRecord(record: SourceIndexRecord): string[] {
  const errors: string[] = [];
  if (!record.id.trim()) errors.push("id is required");
  if (!record.sourceUrl.startsWith("https://")) errors.push("sourceUrl must be HTTPS");
  if (!record.canonicalUrl.startsWith("https://")) errors.push("canonicalUrl must be HTTPS");
  if (!record.title.trim()) errors.push("title is required");
  if (!record.bureau.trim()) errors.push("bureau is required");
  if (!record.contentType.includes("/")) errors.push("contentType must be MIME-like");
  if (!record.firstSeenAt.endsWith("Z")) errors.push("firstSeenAt must end in Z");
  if (!record.lastSeenAt.endsWith("Z")) errors.push("lastSeenAt must end in Z");
  if (record.fetchedAt && !record.fetchedAt.endsWith("Z")) errors.push("fetchedAt must end in Z");

  for (const redirect of record.redirects || []) {
    if (!redirect.fromUrl.startsWith("https://")) errors.push("redirect fromUrl must be HTTPS");
    if (!redirect.toUrl.startsWith("https://")) errors.push("redirect toUrl must be HTTPS");
    if (!REDIRECT_CODES.has(redirect.statusCode)) errors.push("redirect statusCode must be an HTTP redirect code");
    if (!redirect.observedAt.endsWith("Z")) errors.push("redirect observedAt must end in Z");
  }

  if (record.redirects.length > 0) {
    if (record.redirects[0].fromUrl !== record.sourceUrl) errors.push("first redirect must start at sourceUrl");
    if (record.redirects[record.redirects.length - 1].toUrl !== record.canonicalUrl) {
      errors.push("last redirect must end at canonicalUrl");
    }
  } else if (record.sourceUrl !== record.canonicalUrl) {
    errors.push("redirects are required when sourceUrl differs from canonicalUrl");
  }

  if (record.crawlStatus === "fetched" || record.crawlStatus === "not_modified") {
    if (!record.fetchedAt) errors.push(`${record.crawlStatus} records require fetchedAt`);
    if (record.httpStatus !== 200 && record.httpStatus !== 304) {
      errors.push(`${record.crawlStatus} records require httpStatus 200 or 304`);
    }
    if (!record.contentHash || !HASH_RE.test(record.contentHash)) {
      errors.push(`${record.crawlStatus} records require sha256 contentHash`);
    }
  }
  if (record.crawlStatus === "not_modified" && record.httpStatus !== 304) {
    errors.push("not_modified records require httpStatus 304");
  }
  if (record.crawlStatus === "skipped" && !record.skipReason?.trim()) errors.push("skipped records require skipReason");
  if (record.crawlStatus === "failed" && !record.failureReason?.trim()) errors.push("failed records require failureReason");
  return errors;
}
